src/Sampler.cpp

```diff
@@ -0,0 +1,233 @@
+ * availability, see https://www.github.com/llnl/smoothG.
+ *
+ * smoothG is free software; you can redistribute it and/or modify it under the
+ * terms of the GNU Lesser General Public License (as published by the Free
+ * Software Foundation) version 2.1 dated February 1999.
+ *
+ ***********************************************************************EHEADER*/
+
+/** @file
+
+    @brief Contains Sampler classes for getting permeability coefficients.
+*/
+
+#include "Sampler.hpp"
+
+namespace smoothg
+{
+
+NormalDistribution::NormalDistribution(double mean, double stddev, int seed)
+    :
+    generator_(seed),
+    dist_(mean, stddev)
+{
+}
+
+double NormalDistribution::Sample()
+{
+    double out = dist_(generator_);
+    return out;
+}
+
+SimpleSampler::SimpleSampler(int fine_size, int coarse_size)
+    :
+    fine_size_(fine_size), coarse_size_(coarse_size), sample_(-1)
+{
+    fine_.SetSize(fine_size_);
+    coarse_.SetSize(coarse_size_);
+}
+
+void SimpleSampler::NewSample()
+{
+    sample_++;
+}
+
+mfem::Vector& SimpleSampler::GetFineCoefficient()
+{
+    MFEM_ASSERT(sample_ >= 0, "SimpleSampler in wrong state (call NewSample() first)!");
+    fine_ = (1.0 + sample_);
+    return fine_;
+}
+
+mfem::Vector& SimpleSampler::GetCoarseCoefficient()
+{
+    MFEM_ASSERT(sample_ >= 0, "SimpleSampler in wrong state (call NewSample() first)!");
+    coarse_ = (1.0 + sample_);
+    return coarse_;
+}
+
+PDESampler::PDESampler(std::shared_ptr<const Upscale> fvupscale, int fine_vector_size,
+                       int coarse_aggs, int dimension, double cell_volume, double kappa,
+                       int seed)
+    :
+    fvupscale_(fvupscale),
+    normal_distribution_(0.0, 1.0, seed),
+    fine_vector_size_(fine_vector_size),
+    num_coarse_aggs_(coarse_aggs),
+    cell_volume_(cell_volume),
+    current_state_(NO_SAMPLE)
+{
+    Initialize(dimension, kappa);
+}
+
+/**
+   Initialize the PDESampler based on its own, owned FiniteVolumeUpscale object
+*/
+PDESampler::PDESampler(MPI_Comm comm, int dimension,
+                       double cell_volume, double kappa, int seed,
+                       const mfem::SparseMatrix& vertex_edge,
+                       const mfem::Array<int>& partitioning,
+                       const mfem::HypreParMatrix& edge_d_td,
+                       const mfem::SparseMatrix& edge_boundary_att,
+                       const mfem::Array<int>& ess_attr, double spect_tol, int max_evects,
+                       bool dual_target, bool scaled_dual, bool energy_dual,
+                       bool hybridization)
+    :
+    normal_distribution_(0.0, 1.0, seed),
+    fine_vector_size_(vertex_edge.Height()),
+    num_coarse_aggs_(partitioning.Max() + 1),
+    cell_volume_(cell_volume),
+    current_state_(NO_SAMPLE)
+{
+    mfem::SparseMatrix W_block = SparseIdentity(vertex_edge.Height());
+    W_block *= cell_volume_ * kappa * kappa;
+    mfem::Vector one_weight(vertex_edge.Width());
+    one_weight = 1.0;
```

Comment:
If "kappa^2 u - \triangle u = g" is the pde considering here, the weight will need to take geometry (like face area, characteristic length) into account?

Reply:
Yes. The right thing to do is to use `FiniteVolumeMassIntegrator`, but that requires a BilinearForm and a Mesh, and that seems like a lot of machinery to bring into this class, and I am lazy. But probably I should just do it, since this class only makes sense in the finite volume context anyway.

Reply:
Actually, the current method that you implement (if we exclude cell volume and boundary condition as well) can be interpreted as a graph-based sampler through graph Laplacian. Clearly it has its PDE origin in our case, but this extension may be used to apply to other problems (if we eventually encounter one some day Zzzz). 

Cell volume and other geometrical factors can be integrated into the graph through vertex and edge weights. So that seems to be not a big problem.

For boundary condition, apparently it has been an issue in the original PDE-based method, so it may be not too bad to use Neumann condition, which means simply ignore the boundary. In the reference paper, they solve problem on an extended domain, not sure how to do that in graph nicely if we remove the notion of boundary (the equivalent thing should be to only extend the graph at the vertices that is on the "boundary").